---
title: "Plot ESBL cluster MSAs"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, message= FALSE, warning = FALSE)
```

## Introduction

This document provides details of the top 10 largest ESBL contig clusters.
Clusters were mapped to the cluster reference with minimap2 v2.16 and the -asm5
option. SAM files were used to generate a multiple sequence alignemnt then 
nucleotide diversity at each position calculated using PopGenome v2.7.5 in R. 
Alignments were plotted from the generated PAF files with coverage of the 
reference sequence indicated by colour and whether the sequence is mapped on
the forward (+) or reverse (-) strand. Secondary alignments are shown with
dashed outlines. Plots are in order of cluster size from largest to smallest.


```{r load-dat-ting, include = FALSE}
#library(tidyverse)
library(dplyr)
#library(magrittr)
library(here)
library(patchwork)
library(PopGenome)
library(ggplot2)
library(scales)
library(purrr)
library(readr)
library(stringr)

list.files(here("data-raw/review_comment_work/contig_msa"),
           pattern = "paf$",
           full.names = TRUE) -> paffiles
list.files(here("data-raw/review_comment_work/contig_msa"),
           pattern = "paf$") -> paffile_names

map(paffiles, read_tsv, col_select = 1:12,
  col_names =
    c("qname",
      "qlen",
      "qstart",
      "qend",
      "strand",
      "tname",
      "tlen",
      "tstart",
      "tend",
      "nmatch",
      "alen",
      "mapq")
) -> paf_file_list



list.dirs(here("data-raw/review_comment_work/contig_msa/alignments/"),
          recursive = FALSE) -> msa_dirs
map(msa_dirs,
    PopGenome::readData,
    format = "fasta",
    include.unknown = TRUE) -> msa_list

source(here("data-raw/review_comment_work/contig_msa/plot_cluster_msa.R"))


listout <- list()
for (i in 1:length(paffile_names)) {
 return_contig_cluster_plots( paf_file_list[[i]],
                              msa_list[[i]],
                              gsub("\\.paf","", paffile_names[[i]])) ->
  listout[[i]]

}



```


```{r pressure, , fig.height = 10,fig.width = 8}
for (i in order(map_int(paf_file_list, nrow), decreasing = TRUE)) {
  print(listout[[i]] )
}

```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
