---
title: "Analysis scripts: K. pneumoniae complex genomics"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Reproducible analysis scripts}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
# library(dassimKlebs)
library(rstatix)
library(phytools)
library(ggtree)
library(ape)
library(cowplot)
library(ggpubr)
library(dplyr)
library(forcats)
library(tidyr)
library(here)
library(kableExtra)
library(patchwork)
library(scales)
library(viridis)
library(pheatmap)
library(ggplotify)
library(blantyreESBL)
library(stringr)
library(ggnewscale)

sp_dc <- function(x, k) {
  trimws(format(round(x, k), nsmall = k))
}
write_figs <- FALSE

if (write_figs) {
  if (!dir.exists(here("figures"))) {dir.create(here("figures"))}
  if (!dir.exists(here("tables"))) {dir.create(here("tables"))}
  
    if (!dir.exists(here("figures/kleb-genomics"))) {
    dir.create(here("figures/kleb-genomics"))
  }
  
  if (!dir.exists(here("tables/kleb-genomics"))) {
    dir.create(here("tables/kleb-genomics"))
    }
  
}


```

## Introduction

This document generates the tables and figures for the manuscript:

<br />

*Genomic and antigenic diversity of carried _Klebsiella pneumoniae_ isolates mirrors that of invasive isolates in Blantyre, Malawi *

<br />

Joseph M Lewis^1,2,3,4^, , Madalitso Mphasa^1^, Rachel Banda^1^,  Matthew Beale^4^, Jane Mallewa^5^, Eva Heinz^2^, Nicholas Thomson^4,6^, Nicholas A Feasey^1,2^

1. Malawi Liverpool Wellcome Clinical Research Programme, Blantyre, Malawi
2. Department of Clinical Sciences, Liverpool School of Tropical Medicine, Liverpool, UK
3. Department of Clinical Infection, Microbiology and Immunology, University of Liverpool, Liverpool, UK
4. Wellcome Sanger Institute, Hinxton, UK
5. College of Medicine, University of Malawi, Malawi
6. London School of Tropical Medicine and Hygiene, London, UK   

## Assembly stats and accession numbers

```{r assembly_stats}
# load accessions and assembly stats

btESBL_sequence_sample_metadata %>%
  filter(grepl("Klebsiella", species)) %>% 
  transmute("Sample accession" = accession,
         "Lane name" = lane,
         "No. contigs" = number_of_contigs,
         "N50" = N50) %>% 
  kbl(caption = "Accession numbers and assembly statistics for included samples") %>% 
  kable_classic(full_width = FALSE)

```

## MLST, K and O loci

```{r mlst, fig.width=8, fig.height= 11, fig.cap = "FIGURE X: Diversity of Klebsiella chromasomal sequence type (ST, A), K-type (B) and O-type (C). (D) shows cumulative prevalence for ST, K-type and O-types for all isolates in the collection as a function of the number of K-types, O-types, or STs, with each category ordered in size from largest to smallest. (E) and (F) show ST association of O-type and K-type respectively, where area of point is proportional to number of samples in the study. STs with only a single representative in the collection are excluded from plots A, B, C,E, F."}

btESBL_sequence_sample_metadata %>%
  filter(grepl("Klebsiella", species)) %>% 
  rename_with(~ gsub("kleb_", "", .x)) %>% 
  rename_with(~ gsub("k_locus", "K_locus", .x)) %>% 
  rename_with(~ gsub("o_locus", "O_locus", .x))  ->
  dassimKleb_BTKleb.diversity
  
dassimKleb_BTKleb.diversity %>% 
  group_by(ST) %>% 
  mutate(n = n(),
         ST = paste0("ST",ST)) %>% 
  filter(n > 1, ST != "STNovel") %>%  
  ggplot(aes(fct_infreq(ST))) +
  geom_bar(fill = viridis_pal()(4)[3]) +
  coord_flip() + 
  labs(y = "Number", x = element_blank()) +
  theme_bw()  -> a

dassimKleb_BTKleb.diversity %>% 
  mutate(K_locus = case_when(
    K_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ K_locus,
    TRUE ~ "Unknown")) %>% 
  group_by(K_locus) %>% 
  mutate(n = n()) %>% 
  filter(n > 1) %>%  
  ggplot(aes(fct_infreq(K_locus))) +
  geom_bar(fill = viridis_pal()(4)[2]) +
  coord_flip()  +
  labs(y = "Number", x = element_blank()) +
  theme_bw()  -> b

dassimKleb_BTKleb.diversity%>% 
  mutate(O_locus = case_when(
    O_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ O_locus,
    TRUE ~ "Unknown")) %>% 
  group_by(O_locus) %>% 
  mutate(n = n()) %>% 
 # filter(n > 1) %>%  
  ggplot(aes(fct_infreq(O_locus))) +
  geom_bar(fill = viridis_pal()(4)[1]) +
  coord_flip() +
  labs(y = "Number", x = element_blank()) +
  theme_bw()  -> c

# cum K and O serotypes
bind_rows(
  
  dassimKleb_BTKleb.diversity %>%
    group_by(K_locus) %>%
    summarise(n = n()) %>%
    ungroup() %>%
    arrange(desc(n)) %>%
    mutate(
      n_loci = 1:n(),
      cum_n = cumsum(n),
      cum_prop = cum_n / (max(cum_n)+9),
      type = "K-type"
    ) %>% 
    rename(Locus = K_locus),
  
  dassimKleb_BTKleb.diversity %>%
    group_by(O_locus) %>%
      mutate(O_locus = case_when(
    O_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ O_locus,
    TRUE ~ "Unknown")) %>% 
    filter(O_locus != "Unknown") %>% 
    summarise(n = n()) %>%
    ungroup() %>%
    arrange(desc(n)) %>%
    mutate(
      n_loci = 1:n(),
      cum_n = cumsum(n),
      cum_prop = cum_n / (max(cum_n)+6),     # fudge for unknown O tyoe
      type = "O-type"
    ) %>% 
    rename(Locus = O_locus),
  
  dassimKleb_BTKleb.diversity %>%
    filter(ST != "Novel")  %>% 
    group_by(ST) %>%
    summarise(n = n()) %>%
    ungroup() %>%
    arrange(desc(n)) %>%
    mutate(
      n_loci = 1:n(),
      cum_n = cumsum(n),
      cum_prop = cum_n / max(cum_n),
      type = "ST"
    ) %>% 
    rename(Locus = ST)
) %>% 
  mutate(type = factor(type, levels = 
                         c("O-type",
                           "K-type",
                           "ST"))) %>% 
  ggplot(aes(n_loci, cum_prop, color = type)) +
  geom_line() +
  theme_bw() +
  scale_color_manual(values = viridis_pal()(4)[1:3]) +
  labs(x= "Number of K/O-type/ST", y = "Cum. prevalence") + 
  theme(legend.title = element_blank()) -> d

dassimKleb_BTKleb.diversity %>% 
  mutate(O_locus = case_when(
    O_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ O_locus,
    TRUE ~ "Unknown"),
    ST = paste0("ST", ST)) %>% 
  filter(ST != "STNovel") %>% 
    group_by(ST) %>% 
  mutate(n = n()) %>% 
  filter(n > 1) %>% 
  ungroup() %>% 
  mutate(O_locus = fct_infreq(O_locus),
         ST = fct_infreq(ST)) %>% 
  group_by(O_locus, ST) %>% 
  summarise(n = n()) %>% 
  ggplot(aes(ST,O_locus, size = n)) +
  geom_point(color = viridis_pal()(4)[1]) +
  theme_bw() +
  labs(y = "O-type") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.title = element_blank()) -> e

dassimKleb_BTKleb.diversity %>% 
  mutate(K_locus = case_when(
    K_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ K_locus,
    TRUE ~ "Unknown"),
    ST = paste0("ST", ST)) %>% 
  filter(ST != "STNovel") %>% 
  group_by(ST) %>% 
  mutate(n = n()) %>% 
  filter(n > 1) %>% 
  ungroup() %>% 
  mutate(K_locus = fct_infreq(K_locus),
         ST = fct_infreq(ST)) %>% 
  group_by(K_locus, ST) %>% 
  summarise(n = n()) %>% 
  ggplot(aes(ST,K_locus, size = n)) +
  geom_point(color = viridis_pal()(4)[2]) +
  theme_bw() +
  labs(y = "K-type") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.title = element_blank()) +
  scale_size(breaks = c(5,10,15))-> f



((a + b + (c / d)) / e / f) + 
  plot_layout(heights = c(2,1,2))  + 
  plot_annotation(tag_levels = "A") -> stplot
stplot

if (write_figs) {
  ggsave(
    filename = here("figures/kleb-genomics/F1_stplot.svg"),
    plot = stplot,
    width = 8,
    height = 11
    
  )
  ggsave(
    filename = here("figures/kleb-genomics/F1_stplot.pdf"),
    plot = stplot,
    width = 8,
    height = 11
  )
}

 

```

```{r O_type_table}


dassimKleb_BTKleb.diversity %>% 
    mutate(O_locus = case_when(
        O_locus_confidence %in% 
            c("Good",
              "High",             
              "Very high") ~ O_locus,
        TRUE ~ "Unknown")) %>% 
    group_by(O_locus) %>% 
    summarise(n = n()) %>% 
  arrange(desc(n)) %>% 
  mutate(csum = cumsum(n)) %>% 
  mutate(n = paste0(n, " (", sp_dc(100*n/max(csum),1),"%)")) %>% 
  mutate(prop = paste0(csum, " (", sp_dc(100*csum /max(csum),1), "%)")) %>% 
  select(O_locus, n, prop) %>% 
  kbl(caption = "O-types of carriage isolates",
      col.names = c("O-type",
                    "n (%)",
                    "Cumulative n (%)")) %>% 
  kable_classic(full_width = FALSE)
  
  

```

```{r k_type_table}

dassimKleb_BTKleb.diversity %>%
  mutate(K_locus = case_when(
    K_locus_confidence %in%
      c("Good",
        "High",
        "Very high") ~ K_locus,
    TRUE ~ "Unknown"
  )) %>%
  group_by(K_locus) %>%
  summarise(n = n()) %>%
  arrange(desc(n)) %>%
  mutate(csum = cumsum(n)) %>% 
  mutate(n = paste0(n, " (", sp_dc(100 * n / max(csum), 1), "%)")) %>%
  mutate(prop = paste0(csum, " (", sp_dc(100 * csum / max(csum), 1), "%)")) %>%
  select(K_locus, n, prop) %>%
  kbl(caption = "K-types of carriage isolates",
      col.names = c("K-type",
                    "n (%)",
                    "Cumulative n (%)")) %>%
  kable_classic(full_width = FALSE)

```


## Species and population structure

```{r dassim-tree, fig.height = 12, fig.width = 14, fig.cap = "Midpoint-rooted maximum likelihood core gene phylogeny for all samples in collection (A) and restricted to K. pneumoniae sensu stricto (B) showing chromasomal sequence types where black indicates presence and grey absence."}

# what species from kleborate

table(dassimKleb_BTKleb.diversity$species)
dassimKleb_trees.BTcarriage <- btESBL_coregene_tree_kleb


dassimKleb_BTKleb.diversity %>% 
  select(ST, lane) %>% 
  mutate(ST = paste0("ST", ST)) %>% 
  pivot_wider(id_cols = lane ,
              names_from = ST, 
              values_from = ST,
              values_fn = length,
              values_fill = 0) %>% 
  mutate(lane = gsub("#","_", lane)) %>% 
  as.data.frame() -> 
  mlst.onehot


mlst.onehot[, c("lane",
                names(sort(apply(mlst.onehot[-1], 2, sum),
                           decreasing = TRUE)))] -> mlst.onehot

mlst.onehot[-1] <- lapply(mlst.onehot[-1], as.factor)
rownames(mlst.onehot) <- mlst.onehot$lane

dassimKleb_BTKleb.diversity %>% 
  rename(strain = lane) %>% 
  mutate(strain = gsub("#","_", strain)) %>% 
  select(strain, species) %>% 
  as.data.frame() ->
  spec_hm
rownames(spec_hm) <- spec_hm$strain

get_legend(
ggtree(dassimKleb_trees.BTcarriage) %>% 
  gheatmap(
    select(spec_hm, species), 
    width = 0.05, 
    color = NA, 
    font.size = 3, 
    colnames_angle = 90, 
    colnames_position = "top", 
    colnames_offset_y = 10,
    offset = 0.006 ) + 
  theme(legend.position = "bottom", legend.title = element_blank()) +
  ylim(NA, 220) +
  scale_fill_manual(values = viridis(n=5)[c(2:5,1)])
) -> leg

ggtree(dassimKleb_trees.BTcarriage) %>% 
  gheatmap(
    select(spec_hm, species) %>% 
      rename(Species = species), 
    width = 0.05, 
    color = NA, 
    font.size = 3, 
    colnames_angle = 90, 
    colnames_position = "top", 
    colnames_offset_y = 10,
    offset = 0.006 ) %>% 
  gheatmap(
    select(mlst.onehot,-lane),
    width = 3, 
    color = NA, 
    font.size = 3, 
    colnames_angle = 90, 
    colnames_position = "top", 
    colnames_offset_y = 10,
    offset = 0.05 
  ) +
  theme(legend.position = "none") +
  ylim(NA, 220) +
  scale_fill_manual(values = c("lightgrey", "black", viridis(n=5)[c(2:5,1)])) +
  geom_treescale(x = 0.05, y = 180, offset = 2) -> 
  p1

ggtree(treeio::tree_subset(dassimKleb_trees.BTcarriage, 210, levels_back = 0)) %>% 
    gheatmap(
        select(mlst.onehot,-lane),
        width = 2, 
        color = NA, 
        font.size = 3, 
        colnames_angle = 90, 
        colnames_position = "top", 
        colnames_offset_y = 10,
        offset = 0
    ) +  ylim(NA, 205) +
    theme(legend.position = "none") +
    scale_fill_manual(values = c("lightgrey", "black")) +
  geom_treescale(x =0.002, y = 180, offset = 2) -> p2

((p1 + labs(tag = "A")) / leg / (p2 + labs(tag = "B"))) + 
  plot_layout(heights = c(1, 0.1,1)) -> dassim_trees
  
dassim_trees

if (write_figs) {
ggsave(filename = here("figures/kleb-genomics/SUP_F_dassimtrees.pdf"),
       plot = dassim_trees,
       width = 14,
       height = 16
       
)
ggsave(filename = here("figures/kleb-genomics/SUP_F_dassimtrees.svg"),
       plot = dassim_trees,
       width = 14,
       height = 16
)
}

```

## AMR determinents

```{r amr, fig.cap = "FIGURE X: Prevalence of AMR gene grouped by class", fig.height= 10, fig.width= 6}

# add class ro which resistnce is conferred --------------------------------

quinolone <- "Par|Gyr|Par|Qnr|Qep|Nor|GyrA|GyrB|ParC|ParE"
tetracycline <- "Tet"
sulphonamide <- "Sul"
aminoglycoside <- "Str|Aad|Aac|Aph|Rmt|APH"
streptothricin <- "Sat"
macrolide <- "Mph|Mdf|Erm|Ere"
fosfomycin <- "Fos"
chloramphenicol <- "Cat|FloR|Cml"
trimethoprim <- "Dfr"
rifampicin <- "Arr"
ESBL <- "SHV_12"
penicillinase <- "OKP|SCO|LEN|LAP|AmpC|AmpH"
ampc <- "CMY"

btESBL_amrgenes %>% 
  semi_join(dassimKleb_BTKleb.diversity, 
            by = c( "lane")) %>% 
  select(-genus) %>% 
  rename(gene = ref_seq) %>% 
  # add in QRDR mutations
  bind_rows(
    btESBL_qrdr_mutations %>% 
      filter(genus == "K. pneumoniae complex") %>% 
      select(-genus) %>% 
      semi_join(
        btESBL_CARD_qrdr_mutations, 
        by = c("variant" = "variant",
              "gene" =  "gene")
      ) %>% 
      select(gene, lane) %>% 
      unique() %>% 
      mutate(gene =
               gsub("(^.{1})", '\\U\\1',
                    gene,
                    perl = TRUE))
      ) %>% 
  filter(!grepl("MrdA|MefB", gene)) %>% # AmpH|AMPH|AmpC|
  # add in beta-lactamases
  left_join(
    select(btESBL_NCBI_phenotypic_bl, allele_name, class),
    by = c("gene" = "allele_name")) %>% 
  mutate(class = case_when(
    str_detect(gene, quinolone) ~ "Quinolone",
    str_detect(gene, tetracycline) ~ "Tetracycline",
    str_detect(gene, sulphonamide) ~ "Sulphonamide",
    str_detect(gene, aminoglycoside) ~ "Aminoglycoside",
    str_detect(gene, streptothricin) ~ "Streptothricin",
    str_detect(gene, macrolide) ~ "Macrolide",
    str_detect(gene, fosfomycin) ~ "Fosfomycin",
    str_detect(gene, chloramphenicol) ~ "Chloramphenicol",
    str_detect(gene, rifampicin) ~ "Rifampicin",
    str_detect(gene,trimethoprim) ~ "Trimethoprim",
    str_detect(gene,ESBL) ~ "ESBL",
    str_detect(gene,penicillinase) ~ "Penicillinase",
    str_detect(gene,ampc) ~ "AmpC",
    TRUE ~ class
  )) %>% 
  mutate(gene = if_else(gene == "TEM_95",
                        "TEM_1",
                        gene)) ->
  dassimKleb_BTKleb.amr 

# how many genes per sample

dassimKleb_BTKleb.amr %>% 
  group_by(lane) %>% 
  mutate(n = n()) %>% 
  ungroup() %>% 
  select(lane, n) %>% 
  unique() %>% 
  summarise(
    min = min(n),
    lq = quantile(n, 0.25),
    med = median(n),
    uq = quantile(n, 0.75),
    max = max(n)
  )
    

### Plot overall prevalence



dassimKleb_BTKleb.amr %>%
  filter(!grepl("Oqx", gene)) %>%  # remove efflux and intrinsic K pnemo
  group_by(class) %>%              # penicillinase
  mutate(n_class = length(class),
         n_genes_in_class = n_distinct(gene)) %>%
  select(class, n_class, n_genes_in_class) %>%
  unique() %>%
  arrange(n_class) %>%
  ungroup() %>%
  mutate(
    end = cumsum(n_genes_in_class),
    start = lag(end, default = 0),
    textpos = start + 0.5 * (end - start)
  ) -> annotate.df

dassimKleb_BTKleb.amr %>%
  filter(!grepl("Oqx", gene)) %>%
  group_by(class) %>%
  mutate(
    n_class = length(class),
    n_genes_in_class = n_distinct(gene),
    gene = gsub("_", "-", gene)
  ) %>%
  ggplot(aes(fct_reorder(fct_rev(fct_infreq(
    gene
  )), n_class),
  fill = class)) +
  geom_bar() +
  theme_bw() +
  coord_flip(ylim = c(-5, 210),
             clip = "off",
             expand = FALSE) +
  annotate(
    geom = "segment",
    x = annotate.df$start + 0.5 + 0.2,
    xend = annotate.df$end + 0.5 - 0.2,
    y = -60,
    yend = -60
  ) +
  annotate(
    geom = "text",
    y = -70,
    x = annotate.df$textpos,
    label = annotate.df$class,
    size = 3,
    hjust = 1
  ) +
  labs(y = "Number") +
  theme(
    plot.margin = unit(c(0.2, 0.5, 0.2, 4), "cm"),
    axis.title.y  = element_blank(),
    legend.position = "none"
  ) -> amrplot
amrplot

if (write_figs) {
  ggsave(
    filename = here("figures/kleb-genomics/F2_amr_plot.pdf"),
    plot = amrplot,
    width = 6,
    height = 10
  )
  ggsave(
    filename = here("figures/kleb-genomics/F2_amr_plot.svg"),
    plot = amrplot,
    width = 6,
    height = 10
  )
}

```


## Clustering of AMR genes

```{r cluster-amr, fig.width = 7, fig.height= 7, fig.cap = "SUPPLEMENTARY FIGURE X: Jaccard-distance heatmap of presence of ARIBA-identified AMR genes, clustered with a hierarchical clustering algorithm. Several AMR-gene clusters are apparent. Excluded from this matrix are the chromasomally-integrated gyrA, parC (quinolone resistance genes) and the ampH penicillinase."}

# jaccard distance

dassimKleb_BTKleb.amr %>% 
  filter(!grepl("AmpH|Oqx|ParC|GyrA", gene)) %>% 
  select(-class) %>% 
  pivot_wider(id_cols = lane ,
              names_from = gene, 
              values_from = gene,
              values_fn = length,
              values_fill = 0) -> amr.ariba.onehot


names(amr.ariba.onehot) <- gsub("_","-",names(amr.ariba.onehot))

as.data.frame(
  1- as.matrix(
    dist(t(amr.ariba.onehot[-1]), method = "binary")
  )
) -> clust.amr

ggplotify::as.ggplot(
pheatmap(clust.amr, 
         color = viridis_pal()(100),
         fontsize = 6)
) -> amr.heatmap

if (write_figs) {
ggsave(filename = here("figures/kleb-genomics/SUP_F_amr_heatmap.svg"),
       plot = amr.heatmap,
       width =9,
       height = 8
)

ggsave(filename = here("figures/kleb-genomics/SUP_F_amr_heatmap.pdf"),
       plot = amr.heatmap,
       width =9,
       height = 8
)
}

```



```{r malawi-tree-with-amr, fig.width= 13, fig.height = 12, fig.cap = "Presence of ARIBA-identified AMR genes mapped back to phylogeny for (A) KPI isolates only (B) all samples. Some lineage association of AMR genes is apparent."}



dassimKleb_BTKleb.amr %>% 
  filter(!gene %in% c("OqxA", "OqxB", "FosA")) %>% 
  # remove FosA as not in KpI which we are plottingh
  mutate(lane = gsub("#", "_", lane),
         gene = gsub("_","-", gene)) %>% 
  group_by(class) %>% 
  mutate(n_class = length(class)) %>% 
  ungroup() %>% 
  group_by(gene) %>% 
  mutate(n_gene = length(gene)) %>% 
  arrange(desc(n_class), desc(n_gene)) %>% 
  select(-n_class, -n_gene) %>% 
  pivot_wider(id_cols = lane,
              values_from = class,
              names_from = gene) %>% 

  as.data.frame() ->
  amr.ariba.maptotree

dassimKleb_BTKleb.amr %>% 
    filter(!gene %in% c("OqxA", "OqxB", "FosA")) %>% 
    mutate(lane = gsub("#", "_", lane),
           gene = gsub("_","-", gene)) %>% 
    group_by(class) %>% 
    mutate(n_class = length(class)) %>% 
    ungroup() %>% 
    group_by(gene) %>% 
    mutate(n_gene = length(gene)) %>% 
    arrange(desc(n_class), desc(n_gene)) %>% pull(class) %>% 
  unique() -> class_order

rownames(amr.ariba.maptotree) <- amr.ariba.maptotree$lane

colz = hue_pal()(11)
names(colz) <- sort(unique(dassimKleb_BTKleb.amr %>% 
                        #     filter(class != "Fosfomycin") %>% 
                             pull(class)))

ggtree(treeio::tree_subset(dassimKleb_trees.BTcarriage, 210, levels_back = 0)) %>% 
#ggtree(tree) %>% 
    gheatmap(
        select(amr.ariba.maptotree,-lane),
        width = 5, 
        color = NA, 
        font.size = 3, 
        colnames_angle = 90, 
        colnames_position = "top", 
        colnames_offset_y = 0,
        hjust = 0,
        offset = 0
    ) +
  ylim(NA, 210) +
  scale_fill_manual(name = "Class", values = colz, 
                    breaks = class_order,
                    na.translate = FALSE) ->
  malawi_tree_with_amr_kp1


ggtree(dassimKleb_trees.BTcarriage) %>% 
#ggtree(tree) %>% 
    gheatmap(
        select(amr.ariba.maptotree,-lane),
        width = 5, 
        color = NA, 
        font.size = 3, 
        colnames_angle = 90, 
        colnames_position = "top", 
        colnames_offset_y = 0,
        hjust = 0,
        offset = 0
    ) +
  ylim(NA, 220) +
  scale_fill_manual(name = "Class", values = colz, 
                    breaks = class_order,
                    na.translate = FALSE) ->
  malawi_tree_with_amr_all

(malawi_tree_with_amr_kp1 / malawi_tree_with_amr_all) + plot_annotation(tag_levels = "A") + plot_layout(guides = "collect") -> tree_amr_plt

tree_amr_plt

if (write_figs) {
  ggsave(
    filename = here("figures/kleb-genomics/SUP_F_amr_tree.pdf"),
    plot = tree_amr_plt,
    width = 14,
    height = 12
  )
  
  ggsave(
    filename = here("figures/kleb-genomics/SUP_F_amr_tree.svg"),
    plot = tree_amr_plt,
    width = 14,
    height = 12
  )
}

```


## Diversity of Malawian carriage isolates in a global context

``` {r global-tree, fig.height = 12, fig.width = 8}

dassimKleb_trees.global <- btESBL_kleb_globaltree
btESBL_kleb_global_metadata %>% 
  mutate(`Isolate Type` = case_when(
    `Isolate Type` == "Carriage" ~ "Colonising",
    `Isolate Type` == "Infection" ~ "Invasive",
    TRUE ~ `Isolate Type`)) ->
dassimKleb_globalKleb.metadata 

col <-
  c("white",
    "grey30",
    brewer_pal(palette = "Set3")(6))
names(col) <-
  c("0",
    "1",
    "Human",
    "Animal",
    "Environmental",
    "Invasive",
    "Colonising",
    "ESBL") 



(
  (
    (
      (
        ggtree(
          treeio::tree_subset(dassimKleb_trees.global, 734, levels_back = 0),
          size = 0.3
        ) %>%
          gheatmap(
            select(dassimKleb_globalKleb.metadata, Malawi) %>%
              mutate(Malawi =
                       case_when(
                         Malawi == "1" ~ "Malawi",
                         Malawi == "0" ~ "Not Malawi"
                       )),
            width = 0.03,
            color = NA,
            font.size = 4,
            colnames_angle = 90,
            colnames_position = "top",
            colnames_offset_y = 5,
            hjust = 0
          )  +
          scale_fill_manual(
            values = c("Not Malawi" = "white", "Malawi" = "grey30"),
            name = "Country",
            na.translate = FALSE,
            guide = guide_legend(order = 1)
          )  +
          new_scale_fill()
      ) %>%
        gheatmap(
          select(dassimKleb_globalKleb.metadata, `Sample Source`),
          width = 0.03,
          color = NA,
          font.size = 4,
          colnames_angle = 90,
          colnames_position = "top",
          colnames_offset_y = 5,
          hjust = 0,
          offset = 0.0007
        ) +
        scale_fill_manual(
          values =
            c(
              "Human" = brewer_pal(palette = "Set3")(6)[1],
              "Animal" = brewer_pal(palette = "Set3")(6)[2],
              "Environmental" = brewer_pal(palette = "Set3")(6)[3]
            ),
          na.translate = FALSE,
          name = "Sample\nSource",
          guide = guide_legend(order = 2)
        ) +
        new_scale_fill()
    ) %>%
      gheatmap(
        select(dassimKleb_globalKleb.metadata, `Isolate Type`),
        width = 0.03,
        color = NA,
        font.size = 4,
        colnames_angle = 90,
        colnames_position = "top",
        colnames_offset_y = 5,
        hjust = 0,
        offset = 0.0014
      ) +
      scale_fill_manual(
        values =
          c(
            "Invasive" = brewer_pal(palette = "Set3")(6)[4],
            "Colonising" = brewer_pal(palette = "Set3")(6)[5]
          ),
        na.translate = FALSE,
        name = "Isolate\nType",
        guide = guide_legend(order = 3)
      ) +
      new_scale_fill()
  ) %>%   gheatmap(
    select(dassimKleb_globalKleb.metadata, ESBL) %>%
      mutate(ESBL = if_else(ESBL == "ESBL", "Present", NA_character_)),
    width = 0.03,
    color = NA,
    font.size = 4,
    colnames_angle = 90,
    colnames_position = "top",
    colnames_offset_y = 5,
    hjust = 0,
    offset = 0.0021
  ) +
    scale_fill_manual(
      values =
        c("Present" = brewer_pal(palette = "Set3")(6)[6]),
      na.translate = FALSE,
      name = "ESBL",
      guide = guide_legend(order = 4)
    ) +
    new_scale_fill()
) +
  ylim(NA, 680) +
  geom_cladelabel(
    node = 784,
    label = "ST14",
    align = TRUE,
    offset = -0.002
  ) +
  geom_cladelabel(
    node = 827,
    label = "ST15",
    align = TRUE,
    offset = -0.002
  ) +
  geom_cladelabel(
    node = 1020,
    label = "ST340",
    align = TRUE,
    offset = -0.002
  ) +
  geom_cladelabel(
    node = 1066,
    label = "ST307",
    align = TRUE,
    offset = -0.002
  ) +
  geom_treescale(x = 0.001, y = 550, offset = 5) ->
  globaltree_plot
  

globaltree_plot  

if (write_figs) {
  ggsave(
    filename = here("figures/kleb-genomics/F3_globaltree.pdf"),
    plot = globaltree_plot,
    width = 9,
    height = 12
  )
  ggsave(
    filename = here("figures/kleb-genomics/F3_globaltree.svg"),
    plot = globaltree_plot,
    width = 9,
    height = 12
  )
}
```



## Malawian isolates: comparing carriage and infection


```{r malawitree, fig.height = 10, fig.width = 8, fig.cap = "Midpoint rooted core gene maximum likelihood phylogenetic tree of Malawian isolates."}




# 627  = ST 218
(
(
ggtree(treeio::tree_subset(btESBL_kleb_malawi_allisolate_core_gene_tree, 372, levels_back = 0)) %>%
  gheatmap(
    select(dassimKleb_globalKleb.metadata, `Isolate Type`),
    width = 0.03,
    color = NA,
    font.size = 4,
    colnames_angle = 90,
    colnames_position = "top",
    colnames_offset_y = 5,
    hjust = 0,
    offset = 0.001
  ) + 
  scale_fill_manual(
        values =
          c(
            "Invasive" = brewer_pal(palette = "Set3")(6)[4],
            "Colonising" = brewer_pal(palette = "Set3")(6)[5]
          ),
        na.translate = FALSE,
        name = "Isolate\nType",
        guide = guide_legend(order = 2)
      ) +
  new_scale_fill() 
) %>%   gheatmap(
    select(dassimKleb_globalKleb.metadata, ESBL)  %>%
      mutate(ESBL = if_else(ESBL == "ESBL", "Present", NA_character_)),
    width = 0.03,
    color = NA,
    font.size = 4,
    colnames_angle = 90,
    colnames_position = "top",
    colnames_offset_y = 5,
    hjust = 0,
    offset = 0
  )  +
    scale_fill_manual(
      values =
        c("Present" = brewer_pal(palette = "Set3")(6)[6]),
      na.translate = FALSE,
      name = "ESBL",
      guide = guide_legend(order = 1)
    ) +
    new_scale_fill()
) %>%   gheatmap(
    select(dassimKleb_globalKleb.metadata, ybt,clb,iuc,iro,rmpA,rmpA2) %>% 
      mutate(across(everything(), ~ 
                    if_else(.x == "1", "Present", NA_character_))),
    width = 0.18,
    color = "lightgrey",
    font.size = 4,
    colnames_angle = 90,
    colnames_position = "top",
    colnames_offset_y = 5,
    hjust = 0,
    offset = 0.002
  )  + 
  scale_fill_manual(values = c("Present" = "grey30"),
                    name = "Virulence\nGene", na.translate = FALSE,
                    guide = guide_legend(order = 3)) + 
  ylim(NA, 370) +
  geom_cladelabel(node = 599, label = "ST268", align = TRUE, offset = - 0.003) +
  geom_cladelabel(node = 582, label = "ST218", align = TRUE, offset = - 0.003) +
  geom_cladelabel(node = 377, label = "ST14", align = TRUE, offset = - 0.003) +
  geom_cladelabel(node = 413, label = "ST15", align = TRUE, offset = - 0.003) +
  geom_cladelabel(node = 531, label = "ST340", align = TRUE, offset = - 0.003) +
  geom_cladelabel(node = 607, label = "ST307", align = TRUE, offset = - 0.003) +
    geom_treescale(x = 0.001, y = 280, offset = 2) -> malawi_tree_plot_final

malawi_tree_plot_final

if (write_figs) {
ggsave(filename = here("figures/kleb-genomics/F4_malawitree.pdf"),
       plot = malawi_tree_plot_final,
       width =8,
       height = 10
)
ggsave(filename = here("figures/kleb-genomics/F4_malawitree.svg"),
       plot = malawi_tree_plot_final,
       width =8,
       height = 10
)
}


```


```{r inv_vs_carriage, fig.width = 5, fig.height= 5}

dassimKleb_globalKleb.metadata %>% 
  filter(location == "Malawi") %>% 
  mutate(ESBL = if_else(ESBL== "ESBL","ESBL","Not ESBL")) %>% 
  select(`Isolate Type`, ESBL, ybt,clb,iuc,iro,rmpA, rmpA2) %>% 
  pivot_longer(-c(`Isolate Type`, ESBL)) %>% 
  mutate(value = if_else(value == 1, "Present", "Absent")) %>% 
  filter(!is.na(value)) %>%  
  ggplot(aes(name, fill = value)) + geom_bar(position = "fill") +
  facet_grid(~`Isolate Type`) +
  theme_bw() +
  scale_fill_viridis_d() + 
  theme(axis.text.x = element_text(angle = 45, hjust =1),
        legend.title = element_blank()) +
  labs(y = "Proportion", x = "")-> a

dassimKleb_globalKleb.metadata %>% 
  filter(location == "Malawi",
         !is.na(ESBL)) %>% 
  mutate(ESBL = if_else(ESBL== "ESBL","ESBL","Not ESBL")) %>% 
  select(`Isolate Type`, ESBL, ybt,clb,iuc,iro,rmpA, rmpA2) %>% 
  pivot_longer(-c(`Isolate Type`, ESBL)) %>% 
  mutate(value = if_else(value == 1, "Present", "Absent")) %>% 
  filter(!is.na(value)) %>% 
  ggplot(aes(name, fill = value)) + geom_bar(position = "fill") +
  facet_grid(~ESBL) +
  theme_bw() +
  scale_fill_viridis_d() + 
  theme(axis.text.x = element_text(angle = 45, hjust =1),
        legend.title = element_blank()) +
  labs(y = "Proportion", x = "") -> b

(a / b) + 
  plot_layout(guides = "collect") + 
  plot_annotation(tag_levels = "A") -> virplot
virplot


if (write_figs) {
  ggsave(
    filename = here("figures/kleb-genomics/SUPP_FIG_virplot.pdf"),
    plot = virplot,
    width = 5,
    height = 5
  )
  ggsave(
    filename = here("figures/kleb-genomics/SUPP_FIG_virplot.svg"),
    plot = virplot,
    width = 5,
    height = 5
  )
}
```

## O types infection vs carriage

```{r o-type_infn_vs_carriage, fig.height=4, fig.width=10}

dassimKleb_globalKleb.metadata %>%
  filter(study %in% c("cornick", "DASSIM","musciha"),
         !is.na(O_locus)) %>% 
  mutate(O_locus = case_when(
    O_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ O_locus,
    TRUE ~ "Unknown")) %>% 
   mutate(K_locus = case_when(
    K_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ K_locus,
    TRUE ~ "Unknown")) %>% 
  group_by(`Isolate Type`) %>% 
  mutate(total = length(O_locus)) %>% 
  group_by(O_locus, `Isolate Type`) %>% 
  summarise(prop = length(O_locus)/unique(total),
            lci = binom.test(length(O_locus), unique(total))$conf.int[1],
            uci = binom.test(length(O_locus), unique(total))$conf.int[2]) %>% 
  ggplot(aes(fct_rev(fct_reorder(O_locus,prop)), 
             prop, 
             color = `Isolate Type`,
             shape = `Isolate Type`,
             ymin = lci,
             ymax = uci)) +
  geom_point(position = position_dodge(width = 0.5)) +
  geom_errorbar(width = 0, position = position_dodge(width = 0.5)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust= 1)) +
  labs(x = "O-type", y = "Proportion of samples") +
  scale_color_manual(values = hue_pal()(2)[c(2,1)]) ->
  o_type_distn_plot



dassimKleb_globalKleb.metadata %>%
  filter(study %in% c("cornick", "DASSIM","musciha"),
         !is.na(O_locus)) %>% 
  mutate(O_locus = case_when(
    O_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ O_locus,
    TRUE ~ "Unknown")) %>% 
   mutate(K_locus = case_when(
    K_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ K_locus,
    TRUE ~ "Unknown")) %>% 
  mutate(K_locus = as.factor(K_locus),
         `Isolate Type` = as.factor(`Isolate Type`)) %>% 
  count(K_locus, `Isolate Type`, .drop = FALSE) %>% 
  group_by(`Isolate Type`) %>% 
  mutate(total = sum(n)) %>% 
  group_by(K_locus) %>% 
  filter(sum(n) > 2) %>% 
  group_by(K_locus, `Isolate Type`) %>% 
  summarise(prop = n/unique(total),
            lci = binom.test(n, unique(total))$conf.int[1],
            uci = binom.test(n, unique(total))$conf.int[2]) %>% 
  ggplot(aes(fct_rev(fct_reorder(K_locus,prop)), 
             prop, 
             color = `Isolate Type`,
             shape = `Isolate Type`,
             ymin = lci,
             ymax = uci)) +
  geom_point(position = position_dodge(width = 0.5)) +
  geom_errorbar(width = 0, position = position_dodge(width = 0.5)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust= 1)) +
  labs(x = "K-type", y = "Proportion of samples") +
  scale_color_manual(values = hue_pal()(2)[c(2,1)]) ->
  k_type_distn_plot

o_type_distn_plot + k_type_distn_plot +
  plot_annotation(tag_levels = "A") +
  plot_layout(widths = c(1,1.8), guides = 'collect') &
  theme(legend.position = "bottom") -> o_ktype_distn_plot

o_ktype_distn_plot

if (write_figs) {
  ggsave(
    filename = here("figures/kleb-genomics/F5_OK_type_distn.svg"),
    plot =  o_ktype_distn_plot,
    width = 10,
    height = 4
  )
  
  ggsave(
    filename = here("figures/kleb-genomics/F5_OK_type_distn.pdf"),
    plot =  o_ktype_distn_plot,
    width = 10,
    height = 4
  )
}


```

## Testing independence of distributions of K- and O-types

```{r fisher_k_o_types}

dassimKleb_globalKleb.metadata %>%
  filter(study %in% c("cornick", "DASSIM","musciha"),
         !is.na(O_locus)) %>% 
  mutate(O_locus = case_when(
    O_locus_confidence %in% 
      c("Good",
        "High",             
        "Very high") ~ O_locus,
    TRUE ~ "Unknown")) %>% 
  mutate(K_locus = case_when(
    K_locus_confidence %in% 
      c("Good",
        "High",             
        "Very high") ~ K_locus,
    TRUE ~ "Unknown")) -> df


row_wise_fisher_test(
  table(df$K_locus, df$`Isolate Type`), 
  simulate.p.value = TRUE, 
  p.adjust.method = "fdr") %>% 
  select(-c(n,p.adj.signif)) -> 
  tab 

tab %>%
  kbl(
    caption = "Benjamini-Hochberg corrected p values testing equal distribution of K-types across invasive and carriage isolates",
    col.names = c(
      "K-type",
      "Uncorrected p-value",
      "Benjamini-Hochberg corrected p-values"
    )
  ) %>%
  kable_classic(full_width = FALSE) %>%
  row_spec(which(tab$p.adj < 0.05), bold = TRUE)

row_wise_fisher_test(
  table(df$O_locus, df$`Isolate Type`), 
  simulate.p.value = TRUE, 
  p.adjust.method = "fdr") %>% 
  select(-c(n,p.adj.signif)) -> 
  tab 

tab %>% 
  kbl(caption = "Benjamini-Hochberg corrected p values testing equal distribution of O-types across invasive and carriage isolates",
        col.names = c(
      "O-type",
      "Uncorrected p-value",
      "Benjamini-Hochberg corrected p-values"
    )) %>% 
  kable_classic(full_width = FALSE) %>% 
  row_spec(which(tab$p.adj < 0.05), bold = TRUE)


```

## Odds and ends to calculate for manuscript


```{r odds-and-ends}
length(unique(dassimKleb_BTKleb.diversity$ST))

dassimKleb_BTKleb.diversity %>% 
  group_by(ST) %>% 
  summarise(n = n()) %>% 
  ungroup() %>% 
  summarise(median = median(n),
            lqi = quantile(n, 0.25),
            uqi = quantile(n, 0.75),
            n_singletons = sum(n==1))

length(unique(dassimKleb_BTKleb.diversity$K_locus))

dassimKleb_BTKleb.diversity %>% 
  group_by(K_locus) %>% 
  summarise(n = length(unique(ST))) %>% 
  ungroup() %>% 
  summarise(median = median(n),
            lqi = quantile(n, 0.25),
            uqi = quantile(n, 0.75))

length(unique(dassimKleb_BTKleb.diversity$O_locus))

dassimKleb_BTKleb.diversity %>% 
   mutate(O_locus = case_when(
    O_locus_confidence %in% 
           c("Good",
             "High",             
             "Very high") ~ O_locus,
    TRUE ~ "Unknown")) %>% 
  mutate(O_locus = 
           case_when(
             grepl("O1", O_locus) ~ "O1",
             grepl("O2", O_locus) ~ "O2",
             TRUE ~ O_locus)
  ) %>% 
  filter(O_locus != "Unknown") %>% 
  group_by(O_locus) %>% 
  summarise(n = length(unique(ST))) %>% 
  ungroup() %>% 
  summarise(median = median(n),
            lqi = quantile(n, 0.25),
            uqi = quantile(n, 0.75))


dassimKleb_BTKleb.amr %>% 
  filter(!gene %in% c("OqxA", "OqxB", "AmpH")) %>% 
  group_by(lane, class) %>%
  ungroup() %>%  
  select(lane, class) %>% 
  unique() %>% 
  filter(!is.na(class)) %>% 
  group_by(class) %>% 
  summarise(n = n(), prop = n /203) %>% 
  as.data.frame()


btESBL_qrdr_mutations %>% 
    filter(genus == "K. pneumoniae complex") %>% 
    select(-genus) %>% 
    semi_join(
        btESBL_CARD_qrdr_mutations, 
        by = c("variant" = "variant",
               "gene" =  "gene")) %>% 
       unique() %>% count(gene,variant)

dassimKleb_globalKleb.metadata %>%
    filter(location == "Malawi") %>%
    mutate(ESBL = if_else(ESBL == "ESBL", "ESBL", "Not ESBL")) %>%
    select(ybt, clb, iuc, iro, rmpA, rmpA2) %>%
    summarise(across(everything(), ~ sum(as.numeric(.x), na.rm = TRUE)))

# virulence by carriage vs infection

dassimKleb_globalKleb.metadata %>%
  filter(location == "Malawi") %>%
  mutate(ESBL = if_else(ESBL == "ESBL", "ESBL", "Not ESBL")) %>%
  select(`Isolate Type`, ybt, clb, iuc, iro, rmpA, rmpA2) %>%
  group_by(`Isolate Type`) %>%  
  summarise(across(everything(), ~ sum(as.numeric(.x), na.rm = TRUE)))

dassimKleb_globalKleb.metadata %>%
  filter(location == "Malawi") %>%
  group_by(`Isolate Type`) %>% 
  tally()

dassimKleb_globalKleb.metadata %>%
  filter(location == "Malawi") %>%
  group_by(ESBL) %>% 
  tally()

dassimKleb_globalKleb.metadata %>%
  filter(location == "Malawi") %>%
  mutate(ESBL = if_else(ESBL == "ESBL", "ESBL", "Not ESBL")) %>%
  select(ESBL, ybt, clb, iuc, iro, rmpA, rmpA2) %>%
  group_by(ESBL) %>%  
  summarise(across(everything(), ~ sum(as.numeric(.x), na.rm = TRUE)))

dassimKleb_globalKleb.metadata %>% 
  filter(ST %in% c("ST218", "ST268"))

dassimKleb_globalKleb.metadata %>% 
  filter(location == "Malawi") %>%  
  mutate(vir = clb == 1 |
           iuc == 1 | iro == 1 | rmpA == 1 | rmpA2 == 1) %>%  
              filter(ESBL == "ESBL" & vir == TRUE)

# simpsons diversity

dassimKleb_BTKleb.diversity %>% 
    mutate(K_locus = case_when(
        K_locus_confidence %in% 
            c("Good",
              "High",             
              "Very high") ~ K_locus,
        TRUE ~ NA_character_)) %>% 
    mutate(O_locus = case_when(
        O_locus_confidence %in% 
            c("Good",
              "High",             
              "Very high") ~ O_locus,
        TRUE ~ NA_character_)) -> df



```

### Reproducability, session info

```{r sessinfo}

sessionInfo()

```
