<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data and Code Repository For a Study Describing Antimicrobial Resistance in Blantyre, Malawi • blantyreESBL</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Data and Code Repository For a Study Describing Antimicrobial Resistance in Blantyre, Malawi">
<meta property="og:description" content="Data and code repository for a study to undertand antimicrobial resistance in Blantyre, Malawi">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">blantyreESBL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/analysis-ecoli.html">Analysis scripts: E. coli genomics</a>
    </li>
    <li>
      <a href="articles/analysis-kleb.html">Analysis scripts: K. pneumoniae complex genomics</a>
    </li>
    <li>
      <a href="articles/analysis.html">Analysis scripts: longitudinal ESBL carriage dynamics</a>
    </li>
    <li>
      <a href="articles/contig-clusters.html">Longitudinal carriage: validation of contig-cluster approach</a>
    </li>
    <li>
      <a href="articles/fitting-stan-models.html">Fitting and simulating from Stan models</a>
    </li>
    <li>
      <a href="articles/simulations.html">Longitudinal carriage: simulations to explore impact of one colony pick</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="blantyreesbl">blantyreESBL<a class="anchor" aria-label="anchor" href="#blantyreesbl"></a>
</h1></div>
<p><a href="https://zenodo.org/badge/latestdoi/325831306" class="external-link"><img src="https://zenodo.org/badge/325831306.svg" alt="DOI"></a></p>
<!-- badges: start -->
<!-- badges: end -->
<p>The blantyreESBL R package is a repository for data generated as part
of a study into carriage of resistant bacteria in Blantyre, Malawi. It
also includes reproducible analysis script for three manuscripts arising
from the study:</p>
<p><a href="https://joelewis101.github.io/blantyreESBL/articles/analysis.html" class="external-link"><strong>Colonization
dynamics of extended-spectrum beta-lactamase-producing Enterobacterales
in the gut of Malawian adults</strong></a></p>
<p>Joseph M Lewis<sup>1,2,3,4</sup>, , Madalitso Mphasa<sup>1</sup>,
Rachel Banda<sup>1</sup>, Matthew Beale<sup>4</sup>, Eva
Heinz<sup>2</sup>, Jane Mallewa<sup>5</sup>, Christopher
Jewell<sup>6</sup>, Nicholas R Thomson<sup>4,7</sup>, Nicholas A
Feasey<sup>1,2</sup></p>
<p>Paper available at Nature Microbiology <a href="https://doi.org/10.1038/s41564-022-01216-7" class="external-link">here</a></p>
<p><a href="https://joelewis101.github.io/blantyreESBL/articles/analysis-ecoli.html" class="external-link"><strong>Genomic
analysis of extended-spectrum beta-lactamase (ESBL) producing
Escherichia coli colonising adults in Blantyre, Malawi reveals
previously undescribed diversity</strong></a></p>
<p>Joseph M Lewis<sup>1,2,3,4</sup>, , Madalitso Mphasa<sup>1</sup>,
Rachel Banda<sup>1</sup>, Matthew Beale<sup>4</sup>, Jane
Mallewa<sup>5</sup>, Catherine Anscombe<sup>1,2</sup>, Allan
Zuza<sup>1</sup>, Adam P Roberts<sup>2</sup>, Eva Heinz<em><sup>2</sup>,
Nicholas Thomson</em><sup>4,7</sup>, Nicholas A
Feasey*<sup>1,2</sup></p>
<ul>
<li>= contributed equally</li>
</ul>
<p>Preprint available at BioRxiv <a href="https://doi.org/10.1101/2021.10.07.463523" class="external-link">here</a></p>
<p><a href="https://joelewis101.github.io/blantyreESBL/articles/analysis-kleb.html" class="external-link"><strong>Genomic
and antigenic diversity of colonising <em>Klebsiella pneumoniae</em>
isolates mirrors that of invasive isolates in Blantyre,
Malawi</strong></a></p>
<p>Joseph M Lewis<sup>1,2,3,4</sup>, , Madalitso Mphasa<sup>1</sup>,
Rachel Banda<sup>1</sup>, Matthew Beale<sup>4</sup>, Jane
Mallewa<sup>5</sup>, Eva Heinz<sup>2</sup>, Nicholas
Thomson<sup>4,7</sup>, Nicholas A Feasey<sup>1,2</sup></p>
<p>Available at Microbial Genomics <a href="https://doi.org/10.1099/mgen.0.000778" class="external-link">here</a></p>
<div class="section level3">
<h3 id="installing-and-accessing-data">Installing and accessing data<a class="anchor" aria-label="anchor" href="#installing-and-accessing-data"></a>
</h3>
<p>If you just want the data, then all the data to replicate the
analysis are bundled with the package. To install the package from
GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"https://github.com/joelewis101/blantyreESBL"</span><span class="op">)</span></code></pre></div>
<p>The various data objects are described in the <a href="https://joelewis101.github.io/blantyreESBL/" class="external-link">pkgdown</a> site for
this package <a href="https://joelewis101.github.io/blantyreESBL/reference/index.html" class="external-link">here</a>,
and available via R in the usual way
(i.e. <code><a href="reference/btESBL_participants.html">?btESBL_participants</a></code> brings up the definitions for
the <code>btESBL_participants</code> data. They are all lazy loaded so
will be available for use immediately; they all start with
<code>btESBL_</code> to make it easy to choose the one you want using
autocomplete.</p>
</div>
<div class="section level3">
<h3 id="whole-genome-sequence-accession-numbers-and-metadata">Whole genome sequence accession numbers and metadata<a class="anchor" aria-label="anchor" href="#whole-genome-sequence-accession-numbers-and-metadata"></a>
</h3>
<p>Reads from all isolates sequenced as part of this study have been
depositied in the European Nucleotide Archive (ENA); accession numbers
are available in the <code>btESBL_sequence_sample_metadata</code> data
frame, available on installing the package as above.</p>
</div>
<div class="section level3">
<h3 id="analysis-scripts">Analysis scripts<a class="anchor" aria-label="anchor" href="#analysis-scripts"></a>
</h3>
<p>The analysis scripts to reproduce tables and figures for each
manuscript are available as package vignettes; these can be built when
downloading the package by running:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span>
  <span class="st">"https://github.com/joelewis101/blantyreESBL"</span>, 
  build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>, 
  dependencies <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></code></pre></div>
<p>The <code>dependencies = TRUE</code> option will install all the
packages necessary to run the vignette. Building the vignettes may take
some time - you have been warned!</p>
<p>Alternatively the source code for the vignettes are
<code>analysis.Rmd</code> <code>analysis-ecoli.Rmd</code> and
<code>analysis-kleb.Rmd</code>in the <code>vignettes/</code> folder of
the <a href="https://github.com/joelewis101/blantyreESBL" class="external-link">GitHub</a>
repo or the <a href="https://joelewis101.github.io/blantyreESBL/" class="external-link">pkgdown</a> site for
this package has a rendered version of each vignette.</p>
</div>
<div class="section level3">
<h3 id="running-stan-models">Running Stan models<a class="anchor" aria-label="anchor" href="#running-stan-models"></a>
</h3>
<p>The longitudinal modelling paper uses models fit with Stan, the
probabilistic programming language, via the <em>rstan</em> R package.
Unlike the rest if the vignettes (which run the analysis as they are
built) the Stan models are not fit as part of the package vignettes as
they take a long time to fit. The outputs of the models are available as
data objects and a <a href="https://joelewis101.github.io/blantyreESBL/articles/fitting-stan-models.html" class="external-link">vignette</a>
provides instructions on how to fit and simulate from the posterior. The
Stan code is available as .stan files in the package directory - see
vignette for how to locate it.</p>
</div>
<div class="section level3">
<h3 id="author-affiliations">Author affiliations<a class="anchor" aria-label="anchor" href="#author-affiliations"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Malawi Liverpool Wellcome Research Programme, Blantyre, Malawi</li>
<li>Liverpool School of Tropical Medicine, Liverpool, UK</li>
<li>University of Liverpool, Liverpool, UK</li>
<li>Wellcome Sanger Institute, Hinxton, UK</li>
<li>Kamuzu University of Health Sciences, Malawi</li>
<li>University of Lancaster, Lancaster, UK</li>
<li>London School of Tropical Medicine and Hygiene, London, UK</li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing blantyreESBL</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Joseph Lewis <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-3837-5188" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Joseph Lewis.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
