<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ blantyreESBL</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">blantyreESBL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/analysis-ecoli.html">Analysis scripts: E. coli genomics</a>
    </li>
    <li>
      <a href="../articles/analysis-kleb.html">Analysis scripts: K. pneumoniae complex genomics</a>
    </li>
    <li>
      <a href="../articles/analysis.html">Analysis scripts: longitudinal ESBL carriage dynamics</a>
    </li>
    <li>
      <a href="../articles/contig-clusters.html">Longitudinal carriage: validation of contig-cluster approach</a>
    </li>
    <li>
      <a href="../articles/fitting-stan-models.html">Fitting and simulating from Stan models</a>
    </li>
    <li>
      <a href="../articles/simulations.html">Longitudinal carriage: simulations to explore impact of one colony pick</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="participant-data">Participant data <a href="#participant-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Characteristics and exposures of study participants</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="btESBL_participants.html">btESBL_participants</a></code> </p>
        </td>
        <td><p>Data: Characteristics of study participants.</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_exposures.html">btESBL_exposures</a></code> </p>
        </td>
        <td><p>Data: Exposures of included participants</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_stanmodeldata.html">btESBL_stanmodeldata</a></code> </p>
        </td>
        <td><p>Data: Covariate data to pass to Stan longitudinal carriage model in list format</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_modeldata.html">btESBL_modeldata</a></code> </p>
        </td>
        <td><p>Data: Covariate data to pass to Stan longitudinal carriage model in dataframe format</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="stool-culture-data">Stool culture data <a href="#stool-culture-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Results of selective stool culture for ESBL</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="btESBL_stoolESBL.html">btESBL_stoolESBL</a></code> </p>
        </td>
        <td><p>Data: Results of stool culture for ESBL</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_stoolorgs.html">btESBL_stoolorgs</a></code> </p>
        </td>
        <td><p>Data: Species of bacteria isolated from stool culture for ESBL</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_AST.html">btESBL_AST</a></code> </p>
        </td>
        <td><p>Data: Antimicrobial sensitivity testing for a suvet of isolates</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="whole-genome-sequence-metadata-and-analysis">Whole genome sequence metadata and analysis <a href="#whole-genome-sequence-metadata-and-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Accession numbers of isolates and results of sequencing analysis</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="btESBL_sequence_sample_metadata.html">btESBL_sequence_sample_metadata</a></code> </p>
        </td>
        <td><p>Data: Accession numbers and metadata of isolates sequenced for this study</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_amrgenes.html">btESBL_amrgenes</a></code> </p>
        </td>
        <td><p>Data: AMR genes identified in sequenced isolates (excluding QRDR mutations)</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_qrdr_mutations.html">btESBL_qrdr_mutations</a></code> </p>
        </td>
        <td><p>Data: Identified QRDR mutations</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_plasmidreplicons.html">btESBL_plasmidreplicons</a></code> </p>
        </td>
        <td><p>Data: Plasmid replicons identified in sequenced isolates</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_contigclusters.html">btESBL_contigclusters</a></code> </p>
        </td>
        <td><p>Data: Contig cluster membership</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_contigclusters_sensax.html">btESBL_contigclusters_sensax</a></code> </p>
        </td>
        <td><p>Data: Contig cluster assignment in sensitivity analyses</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_contigclusters_msa_blastoutput.html">btESBL_contigclusters_msa_blastoutput</a></code> </p>
        </td>
        <td><p>Data: Contig clusters BLASTed against ISFinder, PLasmidFinder, SRST2 databases</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_contigclusters_msa_alignments.html">btESBL_contigclusters_msa_alignments</a></code> </p>
        </td>
        <td><p>Data: Contig cluster multiple sequence alignments as PopGenome GENOME objects</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_contigclusters_msa_paf_files.html">btESBL_contigclusters_msa_paf_files</a></code> </p>
        </td>
        <td><p>Data: Contig cluster multiple sequence alignment .paf files</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_coregene_tree_esco.html">btESBL_coregene_tree_esco</a></code> </p>
        </td>
        <td><p>Data: E. coli core gene tree</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_coregene_tree_kleb.html">btESBL_coregene_tree_kleb</a></code> </p>
        </td>
        <td><p>Data: K. pneumoniae complex core gene tree</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_snpdists_esco.html">btESBL_snpdists_esco</a></code> </p>
        </td>
        <td><p>Data: Pairwise SNP-distance matrix for all E. coli genomes</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_snpdists_kleb.html">btESBL_snpdists_kleb</a></code> </p>
        </td>
        <td><p>Data: Pairwise SNP-distance matrix for all K. pneumoniae complex genomes</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_NCBI_phenotypic_bl.html">btESBL_NCBI_phenotypic_bl</a></code> </p>
        </td>
        <td><p>Data: NCBI phenotypic beta-lactamase classifications</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_CARD_qrdr_mutations.html">btESBL_CARD_qrdr_mutations</a></code> </p>
        </td>
        <td><p>Data: CARD described QRDR mutations</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="results-of-modelling-and-simulations">Results of modelling and simulations <a href="#results-of-modelling-and-simulations" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Outputs from longituinal modelling of ESBL carriage</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="btESBL_model1posterior.html">btESBL_model1posterior</a></code> </p>
        </td>
        <td><p>Data: Posterior of fitted carriage model (constant antimicrobial effect)</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_model2posterior.html">btESBL_model2posterior</a></code> </p>
        </td>
        <td><p>Data: Posterior of fitted carriage model (decaying antimicrobial effect)</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_model3posterior.html">btESBL_model3posterior</a></code> </p>
        </td>
        <td><p>Data: Posterior of fitted carriage model (seperate ceftriaxone effect)</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_model2simulations.html">btESBL_model2simulations</a></code> </p>
        </td>
        <td><p>Data: Results of simulations from the model posterior</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_model2simulations_2.html">btESBL_model2simulations_2</a></code> </p>
        </td>
        <td><p>Data: Summary of results of simulations from model2 posterior to compare antimicrobials and hospitalisation</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="global-contextual-analyses">Global contextual analyses <a href="#global-contextual-analyses" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Metadata of context genomes and results of clustering study isolates
with global isolates</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="btESBL_ecoli_horesh_metadata.html">btESBL_ecoli_horesh_metadata</a></code> </p>
        </td>
        <td><p>Data: metadata of 10,146 global E. coli genomes from Horesh et al</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_musicha_metadata.html">btESBL_ecoli_musicha_metadata</a></code> </p>
        </td>
        <td><p>Data: Accession numbers and metadata of 97 E. coli genomes from Malawi from Musciha et al</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_global_popPUNK_clusters.html">btESBL_ecoli_global_popPUNK_clusters</a></code> </p>
        </td>
        <td><p>Data: popPUNK cluster assignment of study E. coli isolates using global database</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_globaltree.html">btESBL_ecoli_globaltree</a></code> </p>
        </td>
        <td><p>Data: E.coli global context tree</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_kleb_global_metadata.html">btESBL_kleb_global_metadata</a></code> </p>
        </td>
        <td><p>Data: Accession numbers and metadata of 682 global K. pneumoniae complex genomes</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_kleb_globaltree.html">btESBL_kleb_globaltree</a></code> </p>
        </td>
        <td><p>Data: K. pneumoniae complex global context phylogeny</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_kleb_malawi_allisolate_core_gene_tree.html">btESBL_kleb_malawi_allisolate_core_gene_tree</a></code> </p>
        </td>
        <td><p>Data: K. pneumoniae complex core gene tree of Malawian carriage and invasive isolates</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="global-e-coli-st-contextual-analysis">Global E. coli ST410 contextual analysis <a href="#global-e-coli-st-contextual-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Metadata of global ST410 isolates and results of contextual genomic
analysis</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="btESBL_ecoli_globalst167_tree.html">btESBL_ecoli_globalst167_tree</a></code> </p>
        </td>
        <td><p>Data: E. coli ST167 global phylogeny</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_st167_metadata.html">btESBL_ecoli_st167_metadata</a></code> </p>
        </td>
        <td><p>Data: metadata of global E. coli ST167 isolates</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_st167_amr.html">btESBL_ecoli_st167_amr</a></code> </p>
        </td>
        <td><p>Data: AMR genes identified in global ST167 isolates</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_st167_plasmids.html">btESBL_ecoli_st167_plasmids</a></code> </p>
        </td>
        <td><p>Data: Plasmid replicons identified in global ST167 isolates</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="global-e-coli-st-contextual-analysis">Global E. coli ST167 contextual analysis <a href="#global-e-coli-st-contextual-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Metadata of global ST167 isolates and results of contextual genomic
analysis</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="btESBL_ecoli_globalst410_tree.html">btESBL_ecoli_globalst410_tree</a></code> </p>
        </td>
        <td><p>Data: E. coli ST410 global phylogeny</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_st410_metadata.html">btESBL_ecoli_st410_metadata</a></code> </p>
        </td>
        <td><p>Data: metadata of global E. coli ST410 isolates</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_st410_amr.html">btESBL_ecoli_st410_amr</a></code> </p>
        </td>
        <td><p>Data: AMR genes identified in global ST410 isolates</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_st410_plasmids.html">btESBL_ecoli_st410_plasmids</a></code> </p>
        </td>
        <td><p>Data: Plasmid replicons identified in global ST410 isolates</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="global-e-coli-st-contextual-analysis">Global E. coli ST131 contextual analysis <a href="#global-e-coli-st-contextual-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Metadata of global ST131 isolates and results of contextual genomic
analysis</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="btESBL_ecoli_globalst131_tree.html">btESBL_ecoli_globalst131_tree</a></code> </p>
        </td>
        <td><p>Data: E. coli ST131 global phylogeny</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_st131_metadata.html">btESBL_ecoli_st131_metadata</a></code> </p>
        </td>
        <td><p>Data: metadata of global E. coli ST131 isolates</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_st131_amr.html">btESBL_ecoli_st131_amr</a></code> </p>
        </td>
        <td><p>Data: AMR genes identified in global ST131 isolates</p></td>
      </tr><tr><td>
          <p><code><a href="btESBL_ecoli_st131_plasmids.html">btESBL_ecoli_st131_plasmids</a></code> </p>
        </td>
        <td><p>Data: Plasmid replicons identified in global ST131 isolates</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Joseph Lewis.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

